### ISSUE TITLE
Minion does not retry connection or die when master is disconnected

### ISSUE THREAD

**Issue by kunal-bajpai** [saltstack/salt#44038]
### Description of Issue/Question
As part of failover process, we switch the master CNAME to the new master. When disconnected with the old master, we expect the minions to retry connection to the provided CNAME so it resolves to the new box or die after some timetime. Currently, it seems like the minion stays up even after the connection to the master fails.

Is there an issue open to resolve this?

### Steps to Reproduce Issue
Provide minion with a CNAME to connect to.
switch cname to new master and stop the old master.
Minion does not connect to new master automatically.

### Versions Report
```$ salt-call --versions-report all
Salt Version:
           Salt: 2016.11.7

Dependency Versions:
           cffi: 1.10.0
       cherrypy: 3.2.2
       dateutil: 2.6.0
      docker-py: Not Installed
          gitdb: 0.6.4
      gitpython: git
          ioflo: Not Installed
         Jinja2: 2.9.6
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 1.0.1
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: 2.17
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.13 (default, Apr 19 2017, 02:44:03)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 16.0.2
           RAET: Not Installed
          smmap: 2.0.1
        timelib: Not Installed
        Tornado: 4.5.1
            ZMQ: 4.1.6

System Versions:
           dist: redhat 6.6 Santiago
        machine: x86_64
        release: 2.6.32-504.el6.x86_64
         system: Linux
        version: Red Hat Enterprise Linux Server 6.6 Santiago
```

**Comment by Ch3LL** [comment]
this is a feature request here: https://github.com/saltstack/salt/issues/20275

which looks like a PR has been submitted but still going through test to before merge: https://github.com/saltstack/salt/pull/36180

**Comment by kunal-bajpai** [comment]
@Ch3LL perhaps I muddled the issue explanation while trying to give the motivation behind it.

The PR you shared is to discover masters in the multimaster setup. My issue is that even with single master setup, when the master shuts down, minion does not re-resolve master URI and try to reconnect nor does it die. This becomes a problem when master URI is a single CNAME as the associated A record can change during failover but the minion does not see this.

All I am looking for is that when connection with master is interrupted, minion should re-resolve master URI and reconnect or die.

**Comment by damon-atkins** [comment]
Based on @kunal-bajpai comments. I would suggest it cache successful details of the master just encase DNS or the resolv.conf is broken,  i.e. store the IPv4 or IPv6 address for last resort in a file and use this if the hostname can not be found.

I agree if the connection is dropped it should do a fresh name lookup, rather than use the original information.  And it should do this every time before connecting.

Keep in mine OS do cache name resolution.  So it can take a while after a DNS change for the OS not to use the cached information, even if you have low TTL. OS name resolution cache could return the wrong result for some time.

**Comment by kunal-bajpai** [comment]
@damon-atkins, agreed. For our use-case, a 10 minute window is acceptable as this situation mainly arises during failovers.

**Comment by damon-atkins** [comment]
@kunal-bajpai  you actual want it to retry forever with a retryTime+random(60).  As the last thing you want is having to log into 1000 OS and start the minions by hand.  And all of the minions trying at the same time to re-connect.

In fact it would be good if Salt-Minion waited and then re-executed itself from scratch, picking up any changes in configuration files. Like changing the protocol.

**Comment by kunal-bajpai** [comment]
We already have a process manager that restarts the minion with a jitter when it dies. That is why I am alright with the minion dying when the connection breaks. But of course, having it retry indefinitely as long (as it resolves afresh each time) will be good.

**Comment by kunal-bajpai** [comment]
@Ch3LL can you take another look at this? I think this issue should be reopened.

**Comment by kunal-bajpai** [comment]
@Ch3LL bumping this issue

**Comment by Ch3LL** [comment]
Hmmm looking through the code it looks like that `prep_ip_port method` does get called for both a single master scenario and multi master so I would think this PR would resolve the issue. Any chance you can give the PR a try?

**Comment by kunal-bajpai** [comment]
@Ch3LL a pre-requisite for this kind of retry is that the minion know when connection with the master drops. Is there a mechanism in place which polls if the master is alive? I see `master_alive_interval` is used in multimaster. Does that schedule work in the single-master setup as well?
